_test/test_sqw/worker_4tests_local.m

```diff
@@ -186,44 +198,75 @@
         end
         
-        % send first "running" log message and set-up starting time.
+        % send first "running" log message and set-up starting time. Runs
+        % asynchronously.
         n_steps = je.n_steps;
         mis.do_logging(0,n_steps);
         %
+        je.do_job_completed = false; % wait at barrier if exception in do_job
         while ~je.is_completed()
             je= je.do_job();
             % when its tested, workers are tested in single Matlab
             % session so it will hand up on synchronization
-            je.do_job_completed = true;
             if ~is_tested
                 if DO_LOGGING
-                    fprintf(fh,'Waiting at barrier for all do_job completeon\n');
-                    disp('WORKER_4TESTS: Waiting at barrier for all do_job completeon ')
+                    fprintf(fh,'Waiting at barrier for all do_job completion\n');
+                    disp('WORKER_4TESTS: Waiting at barrier for all do_job completion ')
+                end
+                % explicitly check for cancellation before data reduction
+                is_canceled = je.is_job_canceled();
+                if is_canceled
+                    error('JOB_EXECUTOR:canceled',...
+                        'Job canceled before synchronization after do_job')
                 end
 
-                
                 % when not tested, the synchronization is mandatory
                 je.labBarrier(false); % Wait until all workers finish their
                 %                       job before reducing the data
             end
+            je.do_job_completed = true; % do not wait at barrier if cancellation here
             if DO_LOGGING
                 disp('WORKER_4TESTS: reducing data ')
                 fprintf(fh,'Reducing data\n');
             end
+            % explicitly check for cancellation before data reduction
+            %  the case of cancellation below
+            is_canceled = je.is_job_canceled();
+            if is_canceled
+                error('JOB_EXECUTOR:canceled',...
+                    'Job canceled before starting reduce_data')
+            end
+            je.do_job_completed = false;
             je = je.reduce_data();
         end
-        % Sent final running message. The node 1 waits for other nodes to
-        % send these this kind of messages
+
+        % Sent final running message. Implicitly check for cancellation.
+        % The node 1 waits for other nodes to send these this kind of messages
         mis.do_logging(n_steps,n_steps);
         if ~is_tested
             % stop other nodes until the node 1 finishes to produce the
             % final message
+            if DO_LOGGING
+                disp('WORKER_4TESTS: JE end of task barrier')
+                fprintf(fh,'arriving at JE end of task barrier\n');
+            end
+
             je.labBarrier(false);
+            je.do_job_completed = true;
```

Comment:
is there any route through this code that won't set `do_job_completed` to true?

Reply:
yes, exception